### DISCUSSION TITLE
Async API

### DISCUSSION THREAD

**Discussion by CircuitCM** [sfu-db/connector-x#368] [Ideas]
Would be very interested in a wrapper function to enable async queries, as this is a high performance library it seems fitting.  My own use case is Postgresql with asyncpg, and while that library is great I use Dataframes a lot.  The best way to get a Dataframe is to convert asyncpg's list of tuples to them, as far as I know.  I assume connector-x would be quicker.

**Comment by wangxiaoying** [comment]
Hi @CircuitCM , thanks for the suggestion! Yeah, I think it will be great to provide async API in connectorx. We will look into this, and you are also very welcome to contribute if you want.

**Reply by CircuitCM** [reply]
@wangxiaoying Great! If no one gets to this before me I will be back to contribute when I have time.

**Reply by romanzdk** [reply]
Hello, @wangxiaoying , any update on this, please?

**Reply by wangxiaoying** [reply]
Hi @romanzdk , sorry that we don't really have the resource to work on this part yet. Please feel free to contribute!

**Comment by CircuitCM** [comment]
Following up on this feature, it's very minimal to wrap in a thread executor and probably doesn't need to go in the main library.  Almost no noticeable overhead from the OS thread.

```
data = await asyncio.to_thread(read_sql,*args,**kwargs)
#or
data = await asyncio.to_thread(lambda : read_sql(*args,**kwargs))
```

**Comment by adriangb** [comment]
Would you be able to provide a sans-IO API instead? Kind of how asyncpg uses the pgproto under the hood. It would be great to have a library that knows how to efficiently convert the data but doesn’t force any IO or execution model.
